awx_collection/plugins/modules/tower_inventory_source.py

```diff
@@ -20,333 +20,346 @@
+        group_by=dict(),
+        overwrite=dict(type='bool'),
+        overwrite_vars=dict(type='bool'),
+        custom_virtualenv=dict(type='str'),
+        timeout=dict(type='int'),
+        verbosity=dict(type='int', choices=[0, 1, 2]),
+        update_on_launch=dict(type='bool'),
+        update_cache_timeout=dict(type='int'),
+        source_project=dict(type='str'),
+        update_on_project_update=dict(type='bool'),
         state=dict(choices=['present', 'absent'], default='present'),
     )
 
-    module = TowerModule(argument_spec=argument_spec, supports_check_mode=True)
-
+    # One question here is do we want to end up supporting this within the ansible module itself (i.e. required if, etc)
+    # Or do we want to let the API return issues with "this dosen't support that", etc.
+    #
+    # GUI OPTIONS:
+    # - - - - - - - manual:	file:	scm:	ec2:	gce	azure_rm	vmware	sat	cloudforms	openstack	rhv	tower	custom
+    # credential		?	?	o	o	r	r		r	r	r		r		r	r	o
+    # source_project		?	?	r	-	-	-		-	-	-		-		-	-	-
+    # source_path		?	?	r	-	-	-		-	-	-		-		-	-	-
+    # verbosity			?	?	o	o	o	o		o	o	o		o		o	o	o
+    # overwrite			?	?	o	o	o	o		o	o	o		o		o	o	o
+    # overwrite_vars		?	?	o	o	o	o		o	o	o		o		o	o	o
+    # update_on_launch		?	?	o	o	o	o		o	o	o		o		o	o	o
+    # update_on_project_launch	?	?	o	-	-	-		-	-	-		-		-	-	-
+    # source_regions		?	?	-	o	o	o		-	-	-		-		-	-	-
+    # instance_filters		?	?	-	o	-	-		o	-	-		-		-	o	-
+    # group_by			?	?	-	o	-	-		o	-	-		-		-	-	-
+    # source_vars*		?	?	-	o	-	o		o	o	o		o		-	-	-
+    # environmet vars*		?	?	o	-	-	-		-	-	-		-		-	-	o
+    # source_script		?	?	-	-	-	-		-	-	-		-		-	-	r
+    #
+    # * - source_vars are labeled environment_vars on project and custom sources
+
+    # Create a module for ourselves
+    module = TowerModule(argument_spec=argument_spec,
+                         supports_check_mode=True,
+                         required_if=[
+                             # We don't want to require source if state is present because
+                             # you might be doing an update to an existing source.
+                             # Later on in the code, we will do a test so that if state: present
+                             # and if we don't have an object, we must have source.
+                             ('source', 'scm', ['source_project', 'source_path']),
+                             ('source', 'gce', ['credential']),
+                             ('source', 'azure_rm', ['credential']),
+                             ('source', 'vmware', ['credential']),
+                             ('source', 'satellite6', ['credential']),
+                             ('source', 'cloudforms', ['credential']),
+                             ('source', 'openstack', ['credential']),
+                             ('source', 'rhv', ['credential']),
+                             ('source', 'tower', ['credential']),
+                             ('source', 'custom', ['source_script']),
```

Comment:
I'm okay applying these rules specific to scm versus cloud sources. But I would prefer to have a local variable which lists all the cloud sources in a list, like `['gce', 'azure_rm', ...]`, and then building this list from those. If there are rules unique to certain cloud sources, then that's going in a direction that I do not want to maintain.